### ISSUE TITLE
Rate Limit blocking multiple IPs even if the limit is not reached

### ISSUE THREAD

**Issue by numToStr** [containous/traefik#7129]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

This is just a report of the question which I posted on the forum and didn't get any possible solution. I think it can be a potential issue in the traefik that's why I am creating this bug report.

Forum: https://community.containo.us/t/rate-limit-blocking-multiple-ips-even-if-the-limit-is-not-reached/7144

So, I was playing with the Rate Limit middleware to figure the best values for it. And, I configured the middleware to look for IP in the X-Forwarded-For header. I just bombarded the server with thousands of requests via BlazeMeter, which has, let say, IP1. And then I was trying to reach the server with my local machine which has IP2. But I was getting 429 most of the time on my local machine. I don't know what I did wrong or if this is the expected behavior.

**Server Setup**

ALB --> Traefik --> Nodejs

> Note: Traefik does receive the user's IP in the X-Forwarded-For header. I pasted the log below.

This thing bugging me for a few days. Any help would be appreciated.

### What did you expect to see?

Traefik shouldn't be rate-limiting my local machine as its IP is different from what BlazeMeter has.

### What did you see instead?

Traefik rate limit my local machine even if the limit is not reached.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
v2.2.8
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```yaml
        request-limit:
            rateLimit:
                average: 100
                burst: 50
                period: 1s
                sourceCriterion:
                    requestHost: true
                    ipStrategy:
                        depth: 1
```

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```json
{
    "ClientAddr": "ALB_IP:29864",
    "ClientHost": "CLIENT_IP",
    "ClientPort": "29864",
    "ClientUsername": "-",
    "DownstreamContentSize": 19,
    "DownstreamStatus": 404,
    "Duration": 129795,
    "OriginContentSize": 19,
    "OriginDuration": 15792,
    "OriginStatus": 404,
    "Overhead": 114003,
    "RequestAddr": "example.com",
    "RequestContentSize": 0,
    "RequestCount": 113,
    "RequestHost": "example.com",
    "RequestMethod": "GET",
    "RequestPath": "/api/home",
    "RequestPort": "-",
    "RequestProtocol": "HTTP/1.1",
    "RequestScheme": "https",
    "RetryAttempts": 0,
    "StartLocal": "2020-07-29T13:09:46.926495358Z",
    "StartUTC": "2020-07-29T13:09:46.926495358Z",
    "downstream_Content-Type": "text/plain; charset=utf-8",
    "downstream_X-Content-Type-Options": "nosniff",
    "entryPointName": "websecure",
    "level": "info",
    "msg": "",
    "origin_Content-Type": "text/plain; charset=utf-8",
    "origin_X-Content-Type-Options": "nosniff",
    "request_Accept-Encoding": "gzip",
    "request_Content-Type": "application/json",
    "request_User-Agent": "okhttp/3.8.0",
    "request_X-Forwarded-For": "CLIENT_IP",
    "request_X-Forwarded-Host": "example.com",
    "request_X-Forwarded-Port": "443",
    "request_X-Forwarded-Proto": "https",
    "request_X-Forwarded-Server": "proxy",
    "request_X-Real-Ip": "ALB_IP",
    "time": "2020-07-29T13:09:46Z"
}
```

**Comment by rtribotte** [comment]
Hello @numToStr,

Thanks for your interest in Traefik!

The sourceCriterion strategy could be either the ip or the request host.
If the two strategies are applied in the configuration the middleware is not created.

If you choose to use the ipStrategy, then, in your case, the #1 ip will the client one, your local machine or BlazeMeter IP.
Then, have you checked that both client IPs are different? (for instance if both are behind a NAT, then they have the same IP)

**Comment by numToStr** [comment]
> If the two strategies are applied in the configuration the middleware is not created.

I think , in this case, traefik should throw an error or some kind of warning.

> If you choose to use the ipStrategy, then, in your case, the #1 ip will the client one, your local machine or BlazeMeter IP.

Later after posting this issue, I also tried with only `ipStrategy`, but the issue was still there. And yes, both my local machine and Blazemeter had different IPs.

**Comment by rtribotte** [comment]
Hello @numToStr,

Have you allowed the ALB IP to the trusted IPs (see the [trusted ips section](https://docs.traefik.io/routing/entrypoints/#forwarded-headers) of the documentation) ?

If not, the `X-Forwarded-For` header would not be accepted by Traefik and therefore the computed key for the source bucket in the rate limit middleware is always empty. That would explain that the identified source is always the same.

WDYT ?

**Comment by numToStr** [comment]
Hi @rtribotte, I have `insecure: true` in the entrypoints config. I guess this shouldn't be a problem.

**Comment by rtribotte** [comment]
I can confirm that i have the following error log when activating both strategies:
`ERRO[2020-08-31T13:45:03+02:00] iPStrategy and RequestHost are mutually exclusive  routerName=router@file entryPointName=websecure`

Can you provide a full dump of your configuration (dynamic and static) ?

**Comment by numToStr** [comment]
Yeah Sure.

`traefik.yml`
```yaml
global:
    checkNewVersion: false
    sendAnonymousUsage: false

accessLog:
    bufferingSize: 100

log:
    format: json

providers:
    docker:
        endpoint: unix:///var/run/docker.sock
        exposedByDefault: false
    file:
        directory: /etc/traefik/providers
entrypoints:
    web:
        http:
            redirections:
                entryPoint:
                    to: websecure
                    scheme: https
                    permanent: true
    websecure:
        address: ":443"
        proxyProtocol:
            insecure: true
        forwardedHeaders:
            insecure: true

tls:
    stores:
        default: {}
```
`file.provider.yml`
```yaml
http:
    middlewares:
        for-container:
            chain:
                middlewares:
                    - compress-it
                    - custom-headers
                    - security-headers
                    - cache-headers
                    - cors-headers
                    - request-limit

        https-only:
            redirectScheme:
                scheme: https
                permanent: true

        compress-it:
            compress: {}

        custom-headers:
            headers:
                customResponseHeaders:
                    Server: milkyway

        security-headers:
            headers:
                frameDeny: true
                contentTypeNosniff: true
                referrerPolicy: no-referrer
                browserXssFilter: true
                stsSeconds: 15552000
                forceSTSHeader: true

        cors-headers:
            headers:
                accessControlAllowCredentials: true
                accessControlAllowHeaders:
                    - Origin
                    - Authorization
                    - Content-Type
                accessControlAllowMethods:
                    - OPTION
                    - GET
                    - POST
                    - PUT
                    - PATCH
                accessControlAllowOriginList:
                    - cricuru.com
                accessControlMaxAge: 100
                addVaryHeader: true

        cache-headers:
            headers:
                customResponseHeaders:
                    Cache-Control: "no-cache, max-age=0"

        request-limit:
            rateLimit:
                average: 1000
                burst: 500
                period: 1m
                sourceCriterion:
                    ipStrategy:
                        depth: 1

        replace-with-api:
            replacePathRegex:
                regex: "^/"
                replacement: "/api"
```
I am using `for-container` middleware on my nodejs server.

**Comment by rtribotte** [comment]
Hi @numToStr,

I cannot find the error in your configuration, and also after testing it, it works for me.
Can you paste the output of the rawdata endpoint (/api/rawdata) by activating the api `--api.insecure`  ?

**Comment by numToStr** [comment]
Hi @rtribotte,

Currently, I don't have access to the production server. But I have a similar `docker-compose` on my local machine. So, here is the output for the same.

```json
{
    "routers": {
        "api@internal": {
            "entryPoints": ["traefik"],
            "service": "api@internal",
            "rule": "PathPrefix(`/api`)",
            "priority": 2147483646,
            "status": "enabled",
            "using": ["traefik"]
        },
        "backend@docker": {
            "entryPoints": ["web", "websecure"],
            "middlewares": ["for-container@file"],
            "service": "backend-example",
            "rule": "PathPrefix(`/backend`)",
            "status": "enabled",
            "using": ["web", "websecure"]
        },
        "dashboard@internal": {
            "entryPoints": ["traefik"],
            "middlewares": [
                "dashboard_redirect@internal",
                "dashboard_stripprefix@internal"
            ],
            "service": "dashboard@internal",
            "rule": "PathPrefix(`/`)",
            "priority": 2147483645,
            "status": "enabled",
            "using": ["traefik"]
        },
        "web-to-websecure@internal": {
            "entryPoints": ["web"],
            "middlewares": ["redirect-web-to-websecure@internal"],
            "service": "noop@internal",
            "rule": "HostRegexp(`{host:.+}`)",
            "priority": 2147483647,
            "status": "enabled",
            "using": ["web"]
        }
    },
    "middlewares": {
        "cache-headers@file": {
            "headers": {
                "customResponseHeaders": {
                    "Cache-Control": "no-cache, max-age=0"
                }
            },
            "status": "enabled"
        },
        "compress-it@file": { "compress": {}, "status": "enabled" },
        "cors-headers@file": {
            "headers": {
                "accessControlAllowCredentials": true,
                "accessControlAllowHeaders": [
                    "Origin",
                    "Authorization",
                    "Content-Type"
                ],
                "accessControlAllowMethods": [
                    "OPTION",
                    "GET",
                    "POST",
                    "PUT",
                    "PATCH"
                ],
                "accessControlAllowOriginList": ["example.com"],
                "accessControlMaxAge": 100,
                "addVaryHeader": true
            },
            "status": "enabled"
        },
        "custom-headers@file": {
            "headers": {
                "customResponseHeaders": {
                    "Server": "milkyway"
                }
            },
            "status": "enabled"
        },
        "dashboard_redirect@internal": {
            "redirectRegex": {
                "regex": "^(http:\\/\\/(\\[[\\w:.]+\\]|[\\w\\._-]+)(:\\d+)?)\\/$",
                "replacement": "${1}/dashboard/",
                "permanent": true
            },
            "status": "enabled",
            "usedBy": ["dashboard@internal"]
        },
        "dashboard_stripprefix@internal": {
            "stripPrefix": { "prefixes": ["/dashboard/", "/dashboard"] },
            "status": "enabled",
            "usedBy": ["dashboard@internal"]
        },
        "for-container@file": {
            "chain": {
                "middlewares": [
                    "compress-it@file",
                    "custom-headers@file",
                    "security-headers@file",
                    "cache-headers@file",
                    "cors-headers@file",
                    "request-limit@file"
                ]
            },
            "status": "enabled",
            "usedBy": ["backend@docker"]
        },
        "https-only@file": {
            "redirectScheme": { "scheme": "https", "permanent": true },
            "status": "enabled"
        },
        "redirect-web-to-websecure@internal": {
            "redirectScheme": {
                "scheme": "https",
                "port": "443",
                "permanent": true
            },
            "status": "enabled",
            "usedBy": ["web-to-websecure@internal"]
        },
        "replace-with-api@file": {
            "replacePathRegex": { "regex": "^/", "replacement": "/api" },
            "status": "enabled"
        },
        "request-limit@file": {
            "rateLimit": {
                "average": 1000,
                "period": 60000000000,
                "burst": 500,
                "sourceCriterion": { "ipStrategy": null }
            },
            "status": "enabled"
        },
        "security-headers@file": {
            "headers": {
                "stsSeconds": 15552000,
                "forceSTSHeader": true,
                "frameDeny": true,
                "contentTypeNosniff": true,
                "browserXssFilter": true,
                "referrerPolicy": "no-referrer"
            },
            "status": "enabled"
        }
    },
    "services": {
        "api@internal": { "status": "enabled", "usedBy": ["api@internal"] },
        "backend-example@docker": {
            "loadBalancer": {
                "servers": [{ "url": "http://172.18.0.4:5000" }],
                "passHostHeader": true
            },
            "status": "enabled",
            "usedBy": ["backend@docker"],
            "serverStatus": { "http://172.18.0.4:5000": "UP" }
        },
        "dashboard@internal": {
            "status": "enabled",
            "usedBy": ["dashboard@internal"]
        },
        "noop@internal": {
            "status": "enabled",
            "usedBy": ["web-to-websecure@internal"]
        }
    }
}
```

**Comment by rtribotte** [comment]
Hello @numToStr,

We found the problem and can confirm the bug, it's due to a configuration parsing issue that happen with yaml only.
As you can see, the ipStrategy is null in the rawdata you pasted (`"sourceCriterion": { "ipStrategy": null }`).
We are going to make a fix.

Thanks for your time and contribution to Traefik !

**Comment by numToStr** [comment]
Hi @rtribotte, Thanks for the investigating the issue. Waiting for the new release.

BTW, Traefik is awesome ❤️.

**Comment by rtribotte** [comment]
@numToStr  Just to let you know, in the mean time, a workaround is to use ipStrategy in lowercase `ipstrategy` in the yaml.

**Comment by numToStr** [comment]
@rtribotte Thanks for letting me know.

**Comment by traefiker** [comment]
Closed by #7233.
